corehq/apps/data_interfaces/models.py

```diff
@@ -1125,8 +1157,81 @@ def properties_fit_definition(self, case_properties):
         return all_match or any_match
 
     def when_case_matches(self, case, rule):
+        self._handle_case_duplicate_new(case, rule)
+        result = self._handle_case_duplicate(case, rule)
+
+        return result
+
+    def _handle_case_duplicate_new(self, case, rule, process_updates=False):
+        if not case_matching_rule_exists_in_es(case, rule):
+            # If the case isn't found, any duplicate information is unreliable
+            # Just queue this specific record up manually via resave_case
+            resave_case(rule.domain, case, send_post_save_signal=False)
```

Comment:
I think one issue I have here is that this may indicate an error somewhere, but then potentially masks it by re-saving the case without identifying the underlying reason for it being missing from ES.

Reply:
Unfortunately, this isn't necessarily an error scenario. When we initially populate the rule, for example, we iterate through all the cases currently stored in the database. Because the cases are stored in our database before they are asynchronously processed by elasticsearch, I'd imagine it is fairly common for initial population to run into cases that are not yet in elasticsearch. 

I think my understanding is that, when cases are missing from elasticsearch, either it is due to the scenario above, or it's due to a hardware issue (like elasticsearch was temporarily unavailable) that we would want to automatically correct. I agree with the SNAFU, where we'd like to avoid papering over an issue elsewhere in our system, but that doesn't seem that likely or egregious to me. Say that we have cases now that don't hit elasticsearch as they should -- how likely are we to realize that? And if I was to move forward with this solution, there seems to be a good chance that we'd still have a chance to detect a systematic issue, as the cases that were not part of deduplication domains would still experience the missing data.

Is there a better path forward? Because of the expected nature of this, publishing errors to Sentry seems out.

Reply:
It sounds like you're saying we cannot distinguish between a record missing from ES because it is new vs because some error occurred causing the record to be missing from ES. That seems bad. It would be trivial to at least check the timestamp on the case and log an error to Sentry if it is more than a minute or so (plenty of time for initial save to ES).

> either it is due to the scenario above, or it's due to a hardware issue

I would not rule out another cause: _bug(s) in our code_. We have seen these in the past, and they have been notoriously difficult to track down.

> Because of the expected nature of this, publishing errors to Sentry seems out.

I disagree. I am in favor of suppressing errors if the case is new, determined by whatever heuristic seems best, but still think we should do it for obviously erroneous scenarios.

Reply:
I'm on the fence on this one as well. I'm not sure that this code should serve as a detector of problems from other areas of our codebase. Doing so certainly makes this code more complicated. It also is going to be error-prone on one side or the other. For example, if we set the recency threshold to one minute, then any time elasticsearch gets backed up by more than 1 minute, 'normal' submissions are now considered errors. If we set the threshold to something more generous, then erroneous scenarios don't raise an error.

As I mentioned, adding error detection comes with a complexity cost. How likely is it that logging these events to Sentry pays for that complexity cost? My impression is that we already routinely ignore most Sentry errors. I've had a goal of reducing Sentry output to only actionable items that we'd never want to ignore, and I'm not sure this new Sentry error would meet that criteria.

I think a main sticking point for me is the current situation, without this proposed code. Due to bugs in our code or architecture issues, cases currently may not reach Elasticsearch, and that poses a problem for the reliability of our system. How do we handle that now? It seems strange to me to lean on a mostly unrelated process to be the error-detector here. What happens if this code gets removed in the future? What code becomes responsible for this detection? I like the idea of making errors easier to trace down, but this kind of feels like a Beyonce-rule analogue to me -- if we care about this reliability, then we should put a dedicated system on it.  

To clarify, are you suggesting that we stop processing for cases that aren't in elasticsearch? Or continue processing, but log the error in Sentry?

Reply:
I'm suggesting that we "continue processing, but log the error in Sentry."

I'm also fine with setting the recency threshold closer to 1 hour or possibly even 1 day. I think that would eventually catch most if not all of the ones where a document is erroneously missing from ES.